Remotion/Web/ClientScript/Scripts/SmartPage.ts

```diff
@@ -0,0 +1,1307 @@
+  };
+
+  public SetIsSubmitting (isAsynchronous: boolean, eventTarget: string, eventArgument: string): SubmitState
+  {
+    var submitterElement = this.GetSubmitterOrActiveElement();
+    if (submitterElement != null)
+    {
+      $(submitterElement).addClass('SmartPageSubmitter');
+    }
+
+    $('html').addClass('SmartPageBusy');
+
+    var isAutoPostback = false;
+    if (submitterElement != null)
+    {
+      var tagName = submitterElement.tagName.toLowerCase();
+      if (tagName === 'input')
+      {
+        var type = (submitterElement as HTMLInputElement).type.toLowerCase();
+        isAutoPostback = type !== 'submit' && type !== 'button';
+      }
+      else if (tagName === 'textarea')
+      {
+        isAutoPostback = true;
+      }
+      else if (tagName === 'select')
+      {
+        isAutoPostback = true;
+      }
+    }
+
+    var submitState: SubmitState = {
+      CancelSubmit: false,
+      IsAsynchronous: isAsynchronous,
+      IsAutoPostback: isAutoPostback,
+      Submitter: submitterElement,
+      EventTarget: eventTarget,
+      EventArgument: eventArgument,
+      NextSubmitState : null
+    };
+    if (this._submitState == null)
+    {
+      this._submitState = submitState;
+    }
+    else if (this._submitState.NextSubmitState == null)
+    {
+      if (this._submitState.IsAutoPostback && !isAutoPostback)
+      {
+        this._submitState.NextSubmitState = submitState;
+      }
+      else
+      {
+        // TODO: the other properties are undefined and not null
+        this._submitState.NextSubmitState = { CancelSubmit: true, Submitter: null } as any;
+      }
+    }
+
+    return this._submitState;
+  };
+
+  public ClearIsSubmitting (isAsynchronous: boolean): void
+  {
+    if (isAsynchronous)
+    {
+      if (this._submitState != null && this._submitState.Submitter != null && this._submitState.Submitter.ownerDocument != null)
+      {
+        $(this._submitState.Submitter).removeClass('SmartPageSubmitter');
+      }
+      //setTimeout (function ()
+    //{
+      var html = $ ('html');
+      html.removeClass('SmartPageBusy');
+
+        // Needed in IE8, Firefox 23, Chrome 4
+        // Does not work in Safari 4 for Windows
+        // Does not work in IE10, IE11
+        var cursorBackUp = html.css ('cursor');
+        html.css ('cursor', 'auto !important');
+        //setTimeout (function ()
+        //{
+          html.css ('cursor', cursorBackUp);
+        //}, 0);
+    //}, 0);
+    }
+
+    this._submitState = null;
+  };
+
+  public IsSubmitting(): boolean
+  {
+    return this._submitState != null;
+  };
+
+  public DisableAbortConfirmation(): void
+  {
+    this._isAbortConfirmationEnabled = false;
+  };
+
+  public ShowAbortConfirmation(): boolean
+  {
+    if (this._isAbortConfirmationEnabled && this._abortMessage != null && (!this._isDirtyStateTrackingEnabled || this._isDirty))
```

Comment:
The extra condition should be an assertion inside. Follow-up task could be to start and introduce tuple-types for some of the data in SmartPage.

Reply:
The extra condition is just there to prevent using `!` on `this._abortMessage` since the type checker does not know that `this._isAbortConfirmationEnabled -> this._abortMessage`. But we can use the `!` instead of adding an additional check

Reply:
Yes, please.